src/industry_gui.cpp

```diff
@@ -1178,7 +1180,66 @@ static const NWidgetPart _nested_industry_directory_widgets[] = {
 	EndContainer(),
 };
 
-typedef GUIList<const Industry*> GUIIndustryList;
+typedef GUIList<const Industry*, std::pair<CargoID, CargoID>> GUIIndustryList;
+
+/** Special cargo filter criteria */
+enum CargoFilterSpecialType {
+	CF_ANY  = CT_NO_REFIT,                  ///< Show all industries (i.e. no filtering)
+	CF_NONE = CT_INVALID,                   ///< Show only industries which do not produce/accept cargo
+};
+
+/** Cargo filter functions */
+static bool CDECL CargoFilter(const Industry* const* industry, const std::pair<CargoID, CargoID> cargoes)
+{
+	auto accepted_cargo = cargoes.first;
+	auto produced_cargo = cargoes.second;
+
+	bool accepted_cargo_matches;
+
+	switch (accepted_cargo)
+	{
+		case CF_ANY:
+			accepted_cargo_matches = true;
+			break;
+
+		case CF_NONE:
+			accepted_cargo_matches = std::all_of(std::begin((*industry)->accepts_cargo), std::end((*industry)->accepts_cargo), [](CargoID cargo) {
+				return cargo == CT_INVALID;
+			});
+			break;
+
+		default:
+			accepted_cargo_matches = std::any_of(std::begin((*industry)->accepts_cargo), std::end((*industry)->accepts_cargo), [&](CargoID cargo) {
+				return cargo == accepted_cargo;
+			});
+			break;
+	}
+
+	bool produced_cargo_matches;
+
+	switch (produced_cargo)
+	{
+		case CF_ANY:
+			produced_cargo_matches = true;
+			break;
+
+		case CF_NONE:
+			produced_cargo_matches = std::all_of(std::begin((*industry)->produced_cargo), std::end((*industry)->produced_cargo), [](CargoID cargo) {
+				return cargo == CT_INVALID;
+			});
+			break;
+
+		default:
+			produced_cargo_matches = std::any_of(std::begin((*industry)->produced_cargo), std::end((*industry)->produced_cargo), [&](CargoID cargo) {
```

Comment:
capturing everything is unnecessary, produced_cargo will do

Reply:
I am unsure about what you mean. Maybe:
```suggestion
			produced_cargo_matches = std::find(std::begin((*industry)->produced_cargo), std::end((*industry)->produced_cargo), produced_cargo) != std::end((*industry)->produced_cargo);
```
?

Reply:
i meant `[produced_cargo](CargoID cargo){...}`, but I think this works better in the general sense, actually. No need for lambdas here. I'd probably pull `(*industry)->produced_cargo` into a variable though, to save having to repeat yourself